src/software/geom/algorithms/step_along_perimeter.cpp

```diff
@@ -0,0 +1,76 @@
+#include "software/geom/algorithms/step_along_perimeter.h"
+
+#include <numeric>
+#include <vector>
+
+#include "software/geom/algorithms/collinear.h"
+#include "software/geom/algorithms/distance.h"
+#include "software/geom/segment.h"
+
+
+Point stepAlongPerimeter(const Polygon& polygon, const Point& start,
+                         double travelDistance)
+{
+    if (travelDistance == 0.0)
+    {
+        return start;
+    }
+
+    std::vector<Segment> polygonSegments = polygon.getSegments();
+    std::size_t startSegmentIdx          = 0;
+
+    int i = 0;
+
+    // find initial segment which contains start point
+    for (Segment& segment : polygonSegments)
+    {
+        Point segmentStart = segment.getStart();
+        Point segmentEnd   = segment.getEnd();
+
+        if (collinear(segmentStart, start, segmentEnd))
+        {
+            startSegmentIdx = i;
+            break;
+        }
+        i++;
+    }
+
+    std::size_t segmentIdx = startSegmentIdx;
+
+    bool isClockwise = travelDistance > 0;
+
+    // if travel distance is negative, it can be equal to perimeter - |travelDistance|.
+    // the fmod function is to support wrapping around and negative distance
+    travelDistance = isClockwise
+                         ? travelDistance
+                         : polygon.perimeter() -
+                               std::fmod(std::abs(travelDistance), polygon.perimeter());
```

Comment:
We try to avoid using ternary operators for clarity. Also, I think we should mod positive travel distance as well
```suggestion
    bool isCounterClockwise = travelDistance < 0;
    travelDistance = std::fmod(std::abs(travelDistance), polygon.perimeter());
    if (isCounterClockwise) {
        travelDistance = polygon.perimeter() - travelDistance;
    }
```